### DISCUSSION TITLE
Expose function for config resolving

### DISCUSSION THREAD

**Discussion by RyanClementsHax** [vercel/next.js#40891] [Ideas]
### Describe the feature you'd like to request

Current behavior:
There is no official or clean way to resolve the nextjs config without any side effects

Desired behavior:
Expose a function that resolves the nextjs config and applies all of the defaults and other logic

Use case:
As the developer of [storybook-addon-next](https://github.com/RyanClementsHax/storybook-addon-next), an addon that aims to make nextjs work seamlessly with storybook, I'm finding it really hard to support all of the features of `next/image` well. Many users want to be able to use loaders and other features that require the next config to be resolved. I could replicate this myself, but the logic to resolve the config and apply the defaults changes between versions frequently thus making this method very brittle to nextjs updates.

### Describe the solution you'd like

Something of the following sort that allows me to resolve the final nextjs config within a user's repo.

Perhaps:
```js
import { resolveConfig } from 'next/config'

const resolvedConfig = resolveConfig()
```

It could also accept parameters to help resolve the config like a file path to the config or the raw nextjs config itself

### Describe alternatives you've considered

I have tried writing the logic manually, but this is brittle. This work is shown [here](https://github.com/RyanClementsHax/storybook-addon-next/blob/8ab2858db275462e03e2f9db645b5cfd36669772/src/config/assignDefaults.ts).

**Comment by SukkaW** [comment]
@RyanClementsHax 

Although not documented, I believe Next.js does export a function to load config. Here is an example of its usage:

```ts
const loadConfig = require('next/dist/server/config').default;
// function loadConfig(
//   phase: string,
//   dir: string,
//   customConfig?: object | null
// ): Promise<NextConfigComplete>

const { PHASE_PRODUCTION_BUILD } = require('next/constants');

const nextConfig = await loadConfig(PHASE_PRODUCTION_BUILD, process.cwd());
```

`next/dist/server/config` is used internally by Next.js' jest preset and storybook plugin.

See also the source code of `next-plugin-storybook`: https://github.com/vercel/next.js/tree/canary/packages/next-plugin-storybook

**Reply by RyanClementsHax** [reply]
thanks for pointing that out. I see that function is really close to what I need. Alas, there are some side effects of it that im worried about. For example, it makes a call to `loadWebpackHook` which performs some side effects that are custom to the way nextjs handles webpack, but such side effects aren't desirable in use cases such as mine where something else manages the webpack instance.

https://github.com/vercel/next.js/blob/canary/packages/next/server/config-utils.ts

**Reply by RyanClementsHax** [reply]
In addition, this function isn't part of the public api surface area of nextjs. Part of what I was asking is that nextjs exposes something stable for third parties to rely on.
